### PR TITLE
[MLv2] Lay the foundation for implementing drill-thrus in MLv2

### PR DESCRIPTION
**Pull request by bshepherdson**
There are still some pieces missing, but this has gone far enough to prove the concept. Follow-up PRs can carry this forward.

## Missing parts

- Three drills (`automatic-insights`, `format`, `zoom-in`) are unimplemented.
- A handful of drills will be returned by `available-drill-thrus` but do not have `drill-thru-method` implementations so they can't be applied.
- Two or three drills have extra FE-only changes that need to be captured somewhere else. (Eg. changing the `card.display`)

## Interface

- `available-drill-thrus` returns opaque maps describing the set of applicable drills
- `display-info` on the drills to get an FE-friendly object
- `drill-thru` to apply a drill to a query when clicked.

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx [hunk 1]
```diff
@@ -236,12 +237,27 @@ class Visualization extends PureComponent {
     const question = this._getQuestionForCardCached(metadata, card);
     const mode = this.getMode(this.props.mode, question);
 
-    return mode
```
Comment (camsaul) [inline frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx:239]: should we leave this namespace out for now?

File: src/metabase/lib/drill_thru.cljc [hunk 2]
```diff
@@ -0,0 +1,612 @@
...
+;;; ---------------------------------------- Internals -------------------------------------------
+(defn- structured-query? [query stage-number]
+  (-> (lib.util/query-stage query stage-number)
+      :lib/type
+      (= :mbql.stage/mbql)))
+
+(defn- drill-thru-dispatch [_query _stage-number drill-thru]
+  (:type drill-thru))
+
+(defmulti drill-thru-method
```
Comment (camsaul) [inline src/metabase/lib/drill_thru.cljc:39]: These should use `lib/hierarchy

**Review by camsaul — APPROVED** [review]
Let's just get the BE stuff in so we can start parallelizing on backend work, I would personally vote for splitting off the FE stuff into a separate PR

File: src/metabase/lib/drill_thru.cljc [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+
+(mu/defn pivot-columns-for-type :- [:sequential lib.metadata/ColumnMetadata]
+  "A helper for the FE. Returns all the columns of the given type which can be used to pivot the query."
+  [drill-thru :- [:and ::lib.schema.drill-thru/drill-thru
+                  [:map [:type [:= :drill-thru/pivot]]]]
+   pivot-type :- ::lib.schema.drill-thru/drill-thru-pivot-types]
+  (get-in drill-thru [:pivots pivot-type]))
+
+;;; ----------------------------------------- Sort -----------------------------------------------
+(mu/defn ^:private sort-drill :- [:maybe ::lib.schema.drill-thru/drill-thru]
```
Comment (camsaul) [inline src/metabase/lib/drill_thru.cljc:370]: Is this missing a `drill-thru-method`?

File: src/metabase/lib/drill_thru.cljc (hunk 2 above)
Comment (camsaul) [reply src/metabase/lib/drill_thru.cljc:39]: Also these need `:arglists` metadata

Comment (camsaul) [reply src/metabase/lib/drill_thru.cljc:39]: I fixed this stuff in https://github.com/metabase/metabase/pull/33572 btw, so we can just merge this in and we can fix in a follow-on PR

File: src/metabase/lib/drill_thru.cljc [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+;; TODO: Directly clicking on a table's primary key opens the object detail, with no popup? How does that work?
+
+;;; ---------------------------------------- Internals -------------------------------------------
+(defn- structured-query? [query stage-number]
+  (-> (lib.util/query-stage query stage-number)
+      :lib/type
+      (= :mbql.stage/mbql)))
+
+(defn- drill-thru-dispatch [_query _stage-number drill-thru]
+  (:type drill-thru))
```
Comment (camsaul) [inline src/metabase/lib/drill_thru.cljc:37]: There are tons of implementations of `drill-thru-method` that are written to take > 3 args

File: frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx (hunk 1 above)
Comment (bshepherdson) [reply frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx:239]: Probably yes. It was for my own testing.

**Comment by bshepherdson** [comment]
I've dropped the hacky changes to `Visualization.jsx`, but kept the `frontend/src/metabase-lib` changes. Denis has a downstream branch that relies on those changes and I'm trying to keep the churn down.

**Comment by bshepherdson** [comment]
> > I've dropped the hacky changes to `Visualization.jsx`, but kept the `frontend/src/metabase-lib` changes. Denis has a downstream branch that relies on those changes and I'm trying to keep the churn down.
> 
> Hey @bshepherdson! You can drop all changes in `Visualization.jsx`, I will update my branch, we will put `availableDrillThrus` call to some other place.

I considered all the changes to that file "hacky changes" :stuck_out_tongue_winking_eye: 
`Visualization.jsx` is not part of the diff anymore.

File: frontend/src/metabase/lib/settings.ts [hunk 5]
```diff
@@ -418,5 +418,13 @@ function makeRegexTest(property: string, regex: RegExp) {
 const initValues =
   typeof window !== "undefined" ? _.clone(window.MetabaseBootstrap) : null;
 
+const settings = new MetabaseSettings(initValues);
+
+if (typeof window !== "undefined") {
+  (
+    window as Window & { __metabaseSettings?: MetabaseSettings }
+  ).__metabaseSettings = settings;
```
[the remark this answers is not in the document]
Comment (bshepherdson) [reply frontend/src/metabase/lib/settings.ts:426]: Yes, actually. It's definitely a hack, but we need to put the settings somewhere that MLv2 can get at them. I'm open to a better alternative!

Originally I wanted to include the `MetabaseSettings` with the JS `Metadata` when it's passed to MLv2. But that's in `Question.ts` methods, and `Question`s are created all over the place, so how to get them from the Redux store to every `Question`? It ended up being a mess, and they're global and immutable anyway, so I just put them on `window` and moved on. We could adjust how this works in a follow-up PR though.

**Pull request merged**